packages/tools/src/tools/annotation/planarFreehandROITool/editLoopCommon.ts

```diff
@@ -0,0 +1,278 @@
+      twoClosestDistanceIndexPairs[1].index
+    );
+
+    this.commonEditData.startCrossingIndex = lowestIndex;
+  } else if (editCanvasPoints.length >= 2) {
+    // Check if extending a line back 6 (Proximity) canvas pixels would cross a line.
+
+    // Extend point back 6 canvas pixels from first point.
+    const dir = vec2.create();
+
+    vec2.subtract(dir, editCanvasPoints[1], editCanvasPoints[0]);
+    vec2.normalize(dir, dir);
+
+    const proximity = 6;
+
+    const extendedPoint: Types.Point2 = [
+      editCanvasPoints[0][0] - dir[0] * proximity,
+      editCanvasPoints[0][1] - dir[1] * proximity,
+    ];
+
+    const crossedLineSegmentFromExtendedPoint =
+      getFirstIntersectionWithPolyline(
+        prevCanvasPoints,
+        extendedPoint,
+        editCanvasPoints[0],
+        isClosedContour
+      );
+
+    if (crossedLineSegmentFromExtendedPoint) {
+      // Add points.
+      const pointsToPrepend = [extendedPoint];
+
+      addCanvasPointsToArray(
+        element,
+        pointsToPrepend,
+        editCanvasPoints[0],
+        this.commonData
+      );
+
+      editCanvasPoints.unshift(...pointsToPrepend);
+
+      this.removePointsUpUntilFirstCrossing(isClosedContour);
+
+      this.commonEditData.editIndex = editCanvasPoints.length - 1;
+      this.commonEditData.startCrossingIndex =
+        crossedLineSegmentFromExtendedPoint[0];
+    }
+  }
+}
+
+/**
+ * Removes the points from the `editCanvasPoints` up until the first crossing of
+ * the `prevCanvasPoints`. This is so we can just insert this line segment
+ * into the contour.
+ */
+function removePointsUpUntilFirstCrossing(isClosedContour: boolean): void {
+  const { editCanvasPoints, prevCanvasPoints } = this.commonEditData;
+  let numPointsToRemove = 0;
+
+  for (let i = 0; i < editCanvasPoints.length - 1; i++) {
+    const firstLine = [editCanvasPoints[i], editCanvasPoints[i + 1]];
+
+    const didCrossLine = !!getFirstIntersectionWithPolyline(
+      prevCanvasPoints,
+      firstLine[0],
+      firstLine[1],
+      isClosedContour
+    );
+
+    // Remove last element
+    numPointsToRemove++;
+
+    if (didCrossLine) {
+      break;
+    }
+  }
+
+  // Remove the points
+  editCanvasPoints.splice(0, numPointsToRemove);
+
+  this.commonEditData.editIndex = editCanvasPoints.length - 1;
+}
+
+/**
+ * Returns `true` if the `editCanvasPoints` crosses the `prevCanvasPoints` a
+ * second time.
+ */
+function checkForSecondCrossing(
+  evt: EventTypes.MouseDragEventType,
+  isClosedContour: boolean
+): boolean {
+  const eventDetail = evt.detail;
+  const { currentPoints, lastPoints } = eventDetail;
+  const canvasPos = currentPoints.canvas;
+  const lastCanvasPoint = lastPoints.canvas;
+  const { prevCanvasPoints } = this.commonEditData;
+
+  const crossedLineSegment = getFirstIntersectionWithPolyline(
+    prevCanvasPoints,
+    canvasPos,
+    lastCanvasPoint,
+    isClosedContour
+  );
+
+  if (!crossedLineSegment) {
+    return false;
+  }
+
+  return true;
```

Comment:
maybe I'm wrong, but the function logic is, if it doesn't not have first crossing it has second crossing? what about none?

Reply:
check for second crossing only checks for the second crossing (i.e. end of the edit line). This method is not even called if there is no first crossing yet.

Maybe this is just due to some confusing of the `getFirstIntersectionWithPolyline`, which is a generic helper actually. The helper returns the _first crossing of the lines given_. The parameters given to this helper inside the method are to try to find the second crossing.

I'll add a comment to clarify this. I see how it can be confusing.